src/collector/server/server.go

```diff
@@ -0,0 +1,201 @@
+	// instead of parsing ConfigFile if ConfigFile is not specified.
+	Config config.Configuration
+
+	// InterruptCh is a programmatic interrupt channel to supply to
+	// interrupt and shutdown the server.
+	InterruptCh <-chan error
+}
+
+// Run runs the server programmatically given a filename for the configuration file.
+func Run(runOpts RunOptions) {
+	var cfg config.Configuration
+	if runOpts.ConfigFile != "" {
+		if err := xconfig.LoadFile(&cfg, runOpts.ConfigFile, xconfig.Options{}); err != nil {
+			fmt.Fprintf(os.Stderr, "unable to load %s: %v", runOpts.ConfigFile, err)
+			os.Exit(1)
+		}
+	} else {
+		cfg = runOpts.Config
+	}
+
+	ctx := context.Background()
+	logger, err := cfg.Logging.Build()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "unable to create logger: %v", err)
+		os.Exit(1)
+	}
+	defer logger.Sync()
+
+	scope, closer, err := cfg.Metrics.NewRootScope()
+	if err != nil {
+		logger.Fatal("could not connect to metrics", zap.Error(err))
+	}
+	defer closer.Close()
+
+	instrumentOpts := instrument.NewOptions().
+		SetMetricsScope(scope).
+		SetZapLogger(logger)
+
+	clusterClient, err := cfg.Etcd.NewClient(instrumentOpts)
+	if err != nil {
+		logger.Fatal("could not create etcd client", zap.Error(err))
+	}
+
+	reporter, err := newReporter(cfg.Reporter, clusterClient, instrumentOpts)
+	if err != nil {
+		logger.Fatal("could not create reporter", zap.Error(err))
+	}
+
+	tagEncoderOptions := serialize.NewTagEncoderOptions()
+	tagDecoderOptions := serialize.NewTagDecoderOptions()
+	tagEncoderPoolOptions := pool.NewObjectPoolOptions().
+		SetInstrumentOptions(instrumentOpts.
+			SetMetricsScope(instrumentOpts.MetricsScope().
+				SubScope("tag-encoder-pool")))
+	tagDecoderPoolOptions := pool.NewObjectPoolOptions().
+		SetInstrumentOptions(instrumentOpts.
+			SetMetricsScope(instrumentOpts.MetricsScope().
+				SubScope("tag-decoder-pool")))
+	tagEncoderPool := serialize.NewTagEncoderPool(tagEncoderOptions,
+		tagEncoderPoolOptions)
+	tagEncoderPool.Init()
+	tagDecoderPool := serialize.NewTagDecoderPool(tagDecoderOptions,
+		tagDecoderPoolOptions)
+	tagDecoderPool.Init()
+
+	handler, err := httpd.NewHandler(reporter, tagEncoderPool,
+		tagDecoderPool, instrumentOpts)
+	if err != nil {
+		logger.Fatal("unable to set up handlers", zap.Error(err))
+	}
+
+	if err := handler.RegisterRoutes(); err != nil {
+		logger.Fatal("unable to register routes", zap.Error(err))
+	}
+
+	listenAddress, err := cfg.ListenAddress.Resolve()
+	if err != nil {
+		logger.Fatal("unable to resolve listen address", zap.Error(err))
+	}
+
+	srv := &http.Server{Addr: listenAddress, Handler: handler.Router()}
+	defer func() {
+		logger.Info("closing server")
+		if err := srv.Shutdown(ctx); err != nil {
+			logger.Error("error closing server", zap.Error(err))
+		}
+	}()
+
+	go func() {
+		logger.Info("starting server", zap.String("address", listenAddress))
+		if err := srv.ListenAndServe(); err != nil {
```

Comment:
shouldnt this fatal or something?

Reply:
Sure thing, I'll make sure `err != http.ErrServerClosed` too so we don't fatal in the "normal" sigterm case where we explicitly shut it down ourselves (need to retain clean exit).